iceoryx_hoofs/include/iceoryx_hoofs/log/platform_building_blocks/logger.hpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_PLATFORM_BUILDING_BLOCKS_LOGGER_HPP
+#define IOX_HOOFS_PLATFORM_BUILDING_BLOCKS_LOGGER_HPP
+
+#include "iceoryx_hoofs/log/platform_building_blocks/logcommon.hpp"
+
+#include <atomic>
+#include <cstdint>
+#include <cstring>
+#include <mutex>
+
+namespace iox
+{
+namespace log
+{
+class LogStream;
+} // namespace log
+
+namespace pbb
+{
+template <uint32_t N>
+// NOLINTJUSTIFICATION required for C-style string comparison; safety guaranteed by strncmp
+// NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+inline constexpr bool equalStrings(const char* lhs, const char (&rhs)[N]) noexcept;
+
+/// @brief Tries to get the log level from the 'IOX_LOG_LEVEL' env variable or uses the specified one if the env
+/// variable is not set
+/// @param[in] logLevel is the log level to be used when the env variable is not set
+/// @note The function uses 'getenv' which is not thread safe and can result in undefined behavior when it is called
+/// from multiple threads or the env variable is changed while the function holds a pointer to the data. For this reason
+/// the function should only be used in the startup phase of the application and only in the main thread.
+LogLevel logLevelFromEnvOr(const LogLevel logLevel) noexcept;
+
+template <typename BaseLogger>
+class Logger : public BaseLogger
+{
+  public:
+    friend class log::LogStream;
+
+    inline static constexpr LogLevel minimalLogLevel() noexcept;
+
+    inline static constexpr bool ignoreLogLevel() noexcept;
+
+    inline static Logger& get() noexcept;
+
+    inline static void init(const LogLevel logLevel = logLevelFromEnvOr(LogLevel::INFO)) noexcept;
+
+    inline static void setActiveLogger(Logger* newLogger) noexcept;
```

Comment:
At the moment it is extremely easy to call `setActiveLogger` with a local object and !always! when the program stops we have a dangling pointer when the logger is not reset to the default logger again. So the clean way to do it would be:
```cxx
int main() {
  MyFancyLogger newLogger;
  auto* oldLogger = Logger::get();
  Logger::setActiveLogger(newLogger);
  // to avoid that on destruction someone logs stuff but newLogger is already out of scope we have to call
  Logger::setActiveLogger(oldLogger);
}
```

This could be avoided when `setActiveLogger` returns a guard which resets the active logger to the default logger as soon has it gets out of scope.

Oh btw. this could also happen when my `newLogger` is a static global. When some other global thingy has called `get` before the global static `newLogger` is instantiated then the destruction order is reverted and you have a dangling pointer.

Reply:
I think a key point is that everything we set here has to reside in static memory. I think most problems can be avoided as long as all the loggers to be set are instantiated before the singleton.

In the end it is only a problem if some other singleton accesses the logger but is destroyed later (after main, after the logger). This must be excluded by overall design, but cannot be done by the logger itself.